pkg/utils/nad.go

```diff
@@ -88,6 +94,172 @@ func NewLayer3NetworkConf(conf string) (*Layer3NetworkConf, error) {
 	return networkConf, nil
 }
 
+func OutdateLayer3NetworkConfPerMode(conf string) (string, error) {
+	if conf == "" {
+		return "", nil
+	}
+
+	layer3NetworkConf := Layer3NetworkConf{}
+	if err := json.Unmarshal([]byte(conf), &layer3NetworkConf); err != nil {
+		return "", fmt.Errorf("unmarshal %s failed, error: %w", conf, err)
+	}
+
+	if layer3NetworkConf.Mode == Auto {
+		layer3NetworkConf.Outdated = true
+	} else {
+		// e.g. both vid and route mode are changed, ensure all legacy fields are wiped out
+		layer3NetworkConf.Outdated = false
+	}
+
+	outdatedRoute, err := json.Marshal(layer3NetworkConf)
+	if err != nil {
+		return "", fmt.Errorf("marshal %v failed, error: %w", layer3NetworkConf, err)
+	}
+
+	return string(outdatedRoute), nil
+}
+
+func OutdateNadLayer3NetworkConf(nad *nadv1.NetworkAttachmentDefinition, nadConf *NetConf) (map[string]string, error) {
+	if nad == nil {
+		return nil, nil
+	}
+	annotations := nad.Annotations
+	routeAnnotation := annotations[KeyNetworkRoute]
+	// no route annotation, nothing to do
+	if routeAnnotation == "" {
+		return annotations, nil
+	}
+
+	// both untag and trunk have vlan id 0, just remove route annotation
+	if nadConf != nil && nadConf.Vlan == 0 {
+		delete(annotations, KeyNetworkRoute)
+		return annotations, nil
+	}
+
+	// mark the vlan related route as outdated when it is auto mode
+	newRouteAnnotation, err := OutdateLayer3NetworkConfPerMode(routeAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	annotations[KeyNetworkRoute] = newRouteAnnotation
+
+	return annotations, nil
+}
+
+// caller ensurs this function is only called when oldNad and newNad have same vlan id
+// the first return is not nil only when real change happens
+func OutdateNadLayer3NetworkConfWhenRouteModeChanges(oldNad, newNad *nadv1.NetworkAttachmentDefinition) (map[string]string, error) {
+	if newNad == nil || oldNad == nil {
```

Comment:
is this scenario even possible? 

The only call to this method is during an update of the net-attach-def object